File: templates/examples/core-components/redis/example-sessions.js
/**
 * Redis Session Management Examples
 * 
 * This example demonstrates:
 * - User session storage in Redis
 * - Session expiration and cleanup
 * - Secure session management
 * - Session data persistence
 * - Multi-device session handling
 * 
 * Usage:
 * npm install redis crypto
 * node example-sessions.js
 */

const redis = require('redis');
const crypto = require('crypto');

// Create Redis client
const client = redis.createClient({
    host: process.env.REDIS_HOST || 'localhost',
    port: process.env.REDIS_PORT || 6379,
    password: process.env.REDIS_PASSWORD || undefined,
    db: process.env.REDIS_DB || 0
});

client.on('error', (err) => {
    console.error('Redis Client Error:', err);
});

// ==========================================
// SESSION MANAGER CLASS
// ==========================================

class SessionManager {
    constructor(redisClient) {
        this.client = redisClient;
        this.defaultTTL = 24 * 60 * 60; // 24 hours
        this.sessionPrefix = 'session:';
        this.userSessionsPrefix = 'user_sessions:';
    }

    /**
     * Generate secure session ID
     */
    generateSessionId() {
        return crypto.randomBytes(32).toString('hex');
    }

    /**
     * Create a new session
     */
    async createSession(userId, sessionData = {}, ttl = this.defaultTTL) {
        const sessionId = this.generateSessionId();
        const sessionKey = this.sessionPrefix + sessionId;
        const userSessionsKey = this.userSessionsPrefix + userId;

        const session = {
            id: sessionId,
            userId,
            createdAt: new Date().toISOString(),
            lastActivity: new Date().toISOString(),
            ipAddress: sessionData.ipAddress || null,
            userAgent: sessionData.userAgent || null,
            data: sessionData.data || {},
            isActive: true
        };

        // Use Redis transaction for atomicity
        const multi = this.client.multi();
        
        // Store session data
        multi.hmset(sessionKey, session);
        multi.expire(sessionKey, ttl);
        
        // Add to user's session list
        multi.sadd(userSessionsKey, sessionId);
        multi.expire(userSessionsKey, ttl + 3600); // Sessions list lives longer
        
        await multi.exec();

        console.log(`Session created: ${sessionId} for user: ${userId}`);
        return sessionId;
    }

    /**
     * Get session data
     */
    async getSession(sessionId) {
        const sessionKey = this.sessionPrefix + sessionId;
        
        try {
            const session = await this.client.hgetall(sessionKey);
            
            if (!session || Object.keys(session).length === 0) {
                return null;
            }

            // Parse JSON fields
            session.data = JSON.parse(session.data || '{}');
            session.isActive = session.isActive === 'true';
            
            return session;
        } catch (error) {
            console.error('Error getting session:', error);
            return null;
        }
    }

    /**
     * Update session data
     */
    async updateSession(sessionId, updates) {
        const sessionKey = this.sessionPrefix + sessionId;
        
        // Check if session exists
        const exists = await this.client.exists(sessionKey);
        if (!exists) {
            return false;
        }

        // Update last activity
        updates.lastActivity = new Date().toISOString();
        
        // If updating data, stringify it
        if (updates.data) {
            updates.data = JSON.stringify(updates.data);
        }

        await this.client.hmset(sessionKey, updates);
        console.log(`Session updated: ${sessionId}`);
        return true;
    }

    /**
     * Update session activity (touch)
     */
    async touchSession(sessionId, ttl = this.defaultTTL) {
        const sessionKey = this.sessionPrefix + sessionId;
        
        // Update last activity and reset TTL
        const multi = this.client.multi();
        multi.hset(sessionKey, 'lastActivity', new Date().toISOString());
        multi.expire(sessionKey, ttl);
        
        const results = await multi.exec();
        return results[0][1] > 0; // Returns true if session existed
    }

    /**
     * Destroy session
     */
    async destroySession(sessionId) {
        const sessionKey = this.sessionPrefix + sessionId;
        
        // Get session to find user ID
        const session = await this.getSession(sessionId);
        if (!session) {
            return false;
        }

        const userSessionsKey = this.userSessionsPrefix + session.userId;
        
        // Remove from both places
        const multi = this.client.multi();
        multi.del(sessionKey);
        multi.srem(userSessionsKey, sessionId);
        
        await multi.exec();
        
        console.log(`Session destroyed: ${sessionId}`);
        return true;
    }

    /**
     * Get all sessions for a user
     */
    async getUserSessions(userId) {
        const userSessionsKey = this.userSessionsPrefix + userId;
        const sessionIds = await this.client.smembers(userSessionsKey);
        
        const sessions = [];
        for (const sessionId of sessionIds) {
            const session = await this.getSession(sessionId);
            if (session) {
                sessions.push(session);
            } else {
                // Clean up dead session reference
                await this.client.srem(userSessionsKey, sessionId);
            }
        }
        
        return sessions;
    }

    /**
     * Destroy all sessions for a user
     */
    async destroyUserSessions(userId) {
        const sessions = await this.getUserSessions(userId);
        
        for (const session of sessions) {
            await this.destroySession(session.id);
        }
        
        console.log(`Destroyed ${sessions.length} sessions for user: ${userId}`);
        return sessions.length;
    }

    /**
     * Destroy all sessions except current one
     */
    async destroyOtherSessions(userId, currentSessionId) {
        const sessions = await this.getUserSessions(userId);
        let destroyedCount = 0;
        
        for (const session of sessions) {
            if (session.id !== currentSessionId) {
                await this.destroySession(session.id);
                destroyedCount++;
            }
        }
        
        console.log(`Destroyed ${destroyedCount} other sessions for user: ${userId}`);
        return destroyedCount;
    }

    /**
     * Clean up expired sessions
     */
    async cleanupExpiredSessions() {
        const pattern = this.sessionPrefix + '*';
        const keys = await this.client.keys(pattern);
        let cleanedCount = 0;
        
        for (const key of keys) {
            const ttl = await this.client.ttl(key);
            if (ttl === -1) {
                // Session has no TTL, check if it's valid
                const session = await this.client.hgetall(key);
                if (session && session.lastActivity) {
                    const lastActivity = new Date(session.lastActivity);
                    const now = new Date();
                    const hoursSinceActivity = (now - lastActivity) / (1000 * 60 * 60);
                    
                    if (hoursSinceActivity > 24) {
                        await this.client.del(key);
                        cleanedCount++;
                    }
                }
            }
        }
        
        console.log(`Cleaned up ${cleanedCount} expired sessions`);
        return cleanedCount;
    }

    /**
     * Get session statistics
     */
    async getSessionStats() {
        const pattern = this.sessionPrefix + '*';
        const keys = await this.client.keys(pattern);
        
        const stats = {
            totalSessions: keys.length,
            activeSessions: 0,
            inactiveSessions: 0,
            sessionsByUser: {}
        };
        
        for (const key of keys) {
            const session = await this.client.hgetall(key);
            if (session) {
                if (session.isActive === 'true') {
                    stats.activeSessions++;
                } else {
                    stats.inactiveSessions++;
                }
                
                const userId = session.userId;
                stats.sessionsByUser[userId] = (stats.sessionsByUser[userId] || 0) + 1;
            }
        }
        
        return stats;
    }
}

// ==========================================
// SECURE SESSION MIDDLEWARE
// ==========================================

class SecureSessionManager extends SessionManager {
    constructor(redisClient, secretKey) {
        super(redisClient);
        this.secretKey = secretKey;
    }

    /**
     * Generate signed session cookie
     */
    generateSessionCookie(sessionId) {
        const timestamp = Date.now().toString();
        const data = sessionId + '.' + timestamp;
        const signature = crypto
            .createHmac('sha256', this.secretKey)
            .update(data)
            .digest('hex');
        
        return data + '.' + signature;
    }

    /**
     * Verify signed session cookie
     */
    verifySessionCookie(cookie) {
        const parts = cookie.split('.');
        if (parts.length !== 3) {
            return null;
        }
        
        const [sessionId, timestamp, signature] = parts;
        const data = sessionId + '.' + timestamp;
        const expectedSignature = crypto
            .createHmac('sha256', this.secretKey)
            .update(data)
            .digest('hex');
        
        if (signature !== expectedSignature) {
            return null;
        }
        
        // Check if timestamp is not too old (24 hours)
        const now = Date.now();
        const cookieTime = parseInt(timestamp);
        if (now - cookieTime > 24 * 60 * 60 * 1000) {
            return null;
        }
        
        return sessionId;
    }

    /**
     * Create secure session with IP and user agent validation
     */
    async createSecureSession(userId, ipAddress, userAgent, data = {}) {
        const sessionData = {
            ipAddress,
            userAgent,
            data: {
                ...data,
                securityFingerprint: this.generateSecurityFingerprint(ipAddress, userAgent)
            }
        };
        
        return await this.createSession(userId, sessionData);
    }

    /**
     * Generate security fingerprint
     */
    generateSecurityFingerprint(ipAddress, userAgent) {
        return crypto
            .createHash('sha256')
            .update(ipAddress + userAgent)
            .digest('hex');
    }

    /**
     * Validate session security
     */
    async validateSessionSecurity(sessionId, ipAddress, userAgent) {
        const session = await this.getSession(sessionId);
        if (!session) {
            return false;
        }
        
        // Check IP address (optional - might change for mobile users)
        if (session.ipAddress && session.ipAddress !== ipAddress) {
            console.warn(`IP address mismatch for session: ${sessionId}`);
            // You might want to invalidate the session here
        }
        
        // Check user agent fingerprint
        const currentFingerprint = this.generateSecurityFingerprint(ipAddress, userAgent);
        if (session.data.securityFingerprint !== currentFingerprint) {
            console.warn(`Security fingerprint mismatch for session: ${sessionId}`);
            return false;
        }
        
        return true;
    }
}

// ==========================================
// SESSION STORE FOR EXPRESS.JS
// ==========================================

class RedisSessionStore {
    constructor(redisClient) {
        this.client = redisClient;
        this.prefix = 'sess:';
    }

    /**
     * Get session (Express.js compatible)
     */
    get(sessionId, callback) {
        const key = this.prefix + sessionId;
        this.client.get(key, (err, result) => {
            if (err) {
                return callback(err);
            }
            
            if (!result) {
                return callback(null, null);
            }
            
            try {
                const session = JSON.parse(result);
                callback(null, session);
            } catch (parseErr) {
                callback(parseErr);
            }
        });
    }

    /**
     * Set session (Express.js compatible)
     */
    set(sessionId, session, callback) {
        const key = this.prefix + sessionId;
        const ttl = session.cookie && session.cookie.maxAge 
            ? Math.floor(session.cookie.maxAge / 1000) 
            : 3600;
        
        this.client.setex(key, ttl, JSON.stringify(session), callback);
    }

    /**
     * Destroy session (Express.js compatible)
     */
    destroy(sessionId, callback) {
        const key = this.prefix + sessionId;
        this.client.del(key, callback);
    }
}

// ==========================================
// EXAMPLE USAGE
// ==========================================

async function runSessionExamples() {
    try {
        const sessionManager = new SessionManager(client);
        const secureManager = new SecureSessionManager(client, 'your-secret-key');

        console.log('=== Basic Session Management ===');
        
        // Create session
        const sessionId = await sessionManager.createSession('user123', {
            ipAddress: '192.168.1.100',
            userAgent: 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36',
            data: { 
                preferences: { theme: 'dark', language: 'en' },
                cartItems: [1, 2, 3]
            }
        });
        
        console.log('Created session:', sessionId);
        
        // Get session
        const session = await sessionManager.getSession(sessionId);
        console.log('Retrieved session:', session);
        
        // Update session data
        await sessionManager.updateSession(sessionId, {
            data: JSON.stringify({ 
                preferences: { theme: 'light', language: 'es' },
                cartItems: [1, 2, 3, 4]
            })
        });
        
        console.log('Session updated');
        
        // Touch session (extend TTL)
        await sessionManager.touchSession(sessionId);
        console.log('Session touched');

        console.log('\\n=== Multi-Device Session Management ===');
        
        // Create multiple sessions for same user
        const session2 = await sessionManager.createSession('user123', {
            ipAddress: '10.0.0.50',
            userAgent: 'Mobile Safari',
            data: { device: 'mobile' }
        });
        
        const session3 = await sessionManager.createSession('user123', {
            ipAddress: '172.16.0.10',
            userAgent: 'Chrome Desktop',
            data: { device: 'desktop' }
        });
        
        // Get all user sessions
        const userSessions = await sessionManager.getUserSessions('user123');
        console.log(`User has ${userSessions.length} active sessions`);
        
        // Destroy other sessions except current
        await sessionManager.destroyOtherSessions('user123', sessionId);
        
        console.log('\\n=== Secure Session Management ===');
        
        // Create secure session
        const secureSessionId = await secureManager.createSecureSession(
            'user456',
            '192.168.1.100',
            'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36'
        );
        
        // Generate signed cookie
        const signedCookie = secureManager.generateSessionCookie(secureSessionId);
        console.log('Signed cookie:', signedCookie);
        
        // Verify signed cookie
        const verifiedSessionId = secureManager.verifySessionCookie(signedCookie);
        console.log('Verified session ID:', verifiedSessionId);
        
        // Validate session security
        const isValid = await secureManager.validateSessionSecurity(
            secureSessionId,
            '192.168.1.100',
            'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36'
        );
        console.log('Session security valid:', isValid);

        console.log('\\n=== Session Statistics ===');
        
        const stats = await sessionManager.getSessionStats();
        console.log('Session stats:', stats);
        
        // Cleanup expired sessions
        const cleanedCount = await sessionManager.cleanupExpiredSessions();
        console.log(`Cleaned up ${cleanedCount} expired sessions`);

        console.log('\\n=== Session Cleanup ===');
        
        // Clean up test sessions
        await sessionManager.destroyUserSessions('user123');
        await sessionManager.destroyUserSessions('user456');
        
    } catch (error) {
        console.error('Error in session examples:', error);
    }
}

// ==========================================
// SESSION MONITORING
// ==========================================

class SessionMonitor {
    constructor(sessionManager) {
        this.sessionManager = sessionManager;
    }

    /**
     * Monitor session activity
     */
    async monitorSessions() {
        const stats = await this.sessionManager.getSessionStats();
        
        console.log(`Session Monitor - ${new Date().toISOString()}`);
        console.log(`Total Sessions: ${stats.totalSessions}`);
        console.log(`Active Sessions: ${stats.activeSessions}`);
        console.log(`Inactive Sessions: ${stats.inactiveSessions}`);
        
        // Alert if too many sessions for a user
        for (const [userId, count] of Object.entries(stats.sessionsByUser)) {
            if (count > 5) {
                console.warn(`User ${userId} has ${count} sessions (potential security risk)`);
            }
        }
    }

    /**
     * Schedule regular cleanup
     */
    scheduleCleanup() {
        setInterval(async () => {
            await this.sessionManager.cleanupExpiredSessions();
        }, 60 * 60 * 1000); // Every hour
    }
}

// Run examples when script is executed directly
if (require.main === module) {
    client.on('ready', async () => {
        await runSessionExamples();
        
        // Start monitoring
        const sessionManager = new SessionManager(client);
        const monitor = new SessionMonitor(sessionManager);
        await monitor.monitorSessions();
        
        // Close connection
        client.quit();
    });
}

module.exports = {
    SessionManager,
    SecureSessionManager,
    RedisSessionStore,
    SessionMonitor
};

File: templates/examples/core-components/redis/example-sessions.py
#!/usr/bin/env python3
"""
Redis Session Management Example (Python)

This example demonstrates how to use Redis for session management
with proper serialization, expiration, and session handling.
"""

import os
import sys
import json
import uuid
from typing import Any, Optional, Dict, List
from dataclasses import dataclass, asdict
from datetime import datetime, timedelta
import redis
from dotenv import load_dotenv

# Load environment variables
load_dotenv()

@dataclass
class User:
    """User data model."""
    id: str
    username: str
    email: str
    role: str
    created_at: str
    last_login: Optional[str] = None

@dataclass
class Session:
    """Session data model."""
    session_id: str
    user_id: str
    username: str
    created_at: str
    last_activity: str
    ip_address: Optional[str] = None
    user_agent: Optional[str] = None
    data: Optional[Dict[str, Any]] = None

class SessionManager:
    """Redis-based session management system."""
    
    def __init__(self, 
                 host: str = None,
                 port: int = None,
                 db: int = None,
                 password: str = None,
                 session_ttl: int = 3600):  # 1 hour default
        """Initialize the session manager."""
        
        # Get configuration from environment variables
        self.host = host or os.getenv("REDIS_HOST", "localhost")
        self.port = port or int(os.getenv("REDIS_PORT", "6379"))
        self.db = db or int(os.getenv("REDIS_DB", "0"))
        self.password = password or os.getenv("REDIS_PASSWORD")
        self.session_ttl = session_ttl
        
        # Redis client
        self.client = redis.Redis(
            host=self.host,
            port=self.port,
            db=self.db,
            password=self.password,
            decode_responses=True
        )
        
        # Test connection
        try:
            self.client.ping()
            print(f"✅ Session manager connected to Redis at {self.host}:{self.port}")
        except redis.ConnectionError as e:
            print(f"❌ Failed to connect to Redis: {e}")
            raise
        
        # Key prefixes
        self.SESSION_PREFIX = "session:"
        self.USER_SESSIONS_PREFIX = "user_sessions:"
        self.ACTIVE_SESSIONS_SET = "active_sessions"
    
    def create_session(self, user: User, ip_address: str = None, user_agent: str = None) -> str:
        """Create a new session for a user."""
        try:
            # Generate session ID
            session_id = str(uuid.uuid4())
            
            # Create session object
            session = Session(
                session_id=session_id,
                user_id=user.id,
                username=user.username,
                created_at=datetime.now().isoformat(),
                last_activity=datetime.now().isoformat(),
                ip_address=ip_address,
                user_agent=user_agent,
                data={}
            )
            
            # Store session
            session_key = f"{self.SESSION_PREFIX}{session_id}"
            session_data = asdict(session)
            
            # Set session with TTL
            self.client.setex(session_key, self.session_ttl, json.dumps(session_data))
            
            # Track user sessions
            user_sessions_key = f"{self.USER_SESSIONS_PREFIX}{user.id}"
            self.client.sadd(user_sessions_key, session_id)
            self.client.expire(user_sessions_key, self.session_ttl)
            
            # Add to active sessions set
            self.client.sadd(self.ACTIVE_SESSIONS_SET, session_id)
            
            print(f"✅ Created session {session_id} for user {user.username}")
            return session_id
            
        except Exception as e:
            print(f"❌ Error creating session: {e}")
            return None
    
    def get_session(self, session_id: str) -> Optional[Session]:
        """Get session by ID."""
        try:
            session_key = f"{self.SESSION_PREFIX}{session_id}"
            session_data = self.client.get(session_key)
            
            if not session_data:
                return None
            
            # Parse session data
            session_dict = json.loads(session_data)
            session = Session(**session_dict)
            
            # Update last activity
            session.last_activity = datetime.now().isoformat()
            self.client.setex(session_key, self.session_ttl, json.dumps(asdict(session)))
            
            return session
            
        except Exception as e:
            print(f"❌ Error getting session {session_id}: {e}")
            return None
    
    def update_session(self, session_id: str, data: Dict[str, Any]) -> bool:
        """Update session data."""
        try:
            session = self.get_session(session_id)
            if not session:
                return False
            
            # Update session data
            if session.data:
                session.data.update(data)
            else:
                session.data = data
            
            session.last_activity = datetime.now().isoformat()
            
            # Save updated session
            session_key = f"{self.SESSION_PREFIX}{session_id}"
            self.client.setex(session_key, self.session_ttl, json.dumps(asdict(session)))
            
            return True
            
        except Exception as e:
            print(f"❌ Error updating session {session_id}: {e}")
            return False
    
    def delete_session(self, session_id: str) -> bool:
        """Delete a session."""
        try:
            session_key = f"{self.SESSION_PREFIX}{session_id}"
            
            # Get session first to find user ID
            session = self.get_session(session_id)
            if session:
                # Remove from user sessions
                user_sessions_key = f"{self.USER_SESSIONS_PREFIX}{session.user_id}"
                self.client.srem(user_sessions_key, session_id)
            
            # Remove from active sessions
            self.client.srem(self.ACTIVE_SESSIONS_SET, session_id)
            
            # Delete session
            result = self.client.delete(session_key)
            
            if result:
                print(f"✅ Deleted session {session_id}")
            
            return bool(result)
            
        except Exception as e:
            print(f"❌ Error deleting session {session_id}: {e}")
            return False
    
    def get_user_sessions(self, user_id: str) -> List[Session]:
        """Get all sessions for a user."""
        try:
            user_sessions_key = f"{self.USER_SESSIONS_PREFIX}{user_id}"
            session_ids = self.client.smembers(user_sessions_key)
            
            sessions = []
            for session_id in session_ids:
                session = self.get_session(session_id)
                if session:
                    sessions.append(session)
            
            return sessions
            
        except Exception as e:
            print(f"❌ Error getting user sessions for {user_id}: {e}")
            return []
    
    def delete_user_sessions(self, user_id: str) -> int:
        """Delete all sessions for a user."""
        try:
            sessions = self.get_user_sessions(user_id)
            deleted_count = 0
            
            for session in sessions:
                if self.delete_session(session.session_id):
                    deleted_count += 1
            
            # Clean up user sessions key
            user_sessions_key = f"{self.USER_SESSIONS_PREFIX}{user_id}"
            self.client.delete(user_sessions_key)
            
            print(f"✅ Deleted {deleted_count} sessions for user {user_id}")
            return deleted_count
            
        except Exception as e:
            print(f"❌ Error deleting user sessions for {user_id}: {e}")
            return 0
    
    def get_active_sessions_count(self) -> int:
        """Get count of active sessions."""
        try:
            return self.client.scard(self.ACTIVE_SESSIONS_SET)
        except Exception as e:
            print(f"❌ Error getting active sessions count: {e}")
            return 0
    
    def cleanup_expired_sessions(self) -> int:
        """Clean up expired sessions from tracking sets."""
        try:
            active_session_ids = self.client.smembers(self.ACTIVE_SESSIONS_SET)
            expired_count = 0
            
            for session_id in active_session_ids:
                session_key = f"{self.SESSION_PREFIX}{session_id}"
                if not self.client.exists(session_key):
                    # Session expired, remove from tracking
                    self.client.srem(self.ACTIVE_SESSIONS_SET, session_id)
                    expired_count += 1
            
            print(f"🧹 Cleaned up {expired_count} expired sessions")
            return expired_count
            
        except Exception as e:
            print(f"❌ Error cleaning up expired sessions: {e}")
            return 0
    
    def get_session_stats(self) -> Dict[str, Any]:
        """Get session statistics."""
        try:
            stats = {
                "active_sessions": self.get_active_sessions_count(),
                "session_ttl": self.session_ttl,
                "redis_info": {
                    "used_memory": self.client.info()["used_memory_human"],
                    "connected_clients": self.client.info()["connected_clients"]
                }
            }
            
            return stats
            
        except Exception as e:
            print(f"❌ Error getting session stats: {e}")
            return {}

class SessionDemo:
    """Demonstration of Redis session management."""
    
    def __init__(self):
        """Initialize the session demo."""
        self.session_manager = SessionManager()
        
        # Sample users
        self.users = {
            "user1": User(
                id="user1",
                username="alice",
                email="alice@example.com",
                role="admin",
                created_at=datetime.now().isoformat()
            ),
            "user2": User(
                id="user2",
                username="bob",
                email="bob@example.com",
                role="user",
                created_at=datetime.now().isoformat()
            ),
            "user3": User(
                id="user3",
                username="charlie",
                email="charlie@example.com",
                role="user",
                created_at=datetime.now().isoformat()
            )
        }
        
        self.active_sessions = {}
    
    def login_user(self, username: str, ip_address: str = "127.0.0.1") -> Optional[str]:
        """Simulate user login."""
        user = None
        for u in self.users.values():
            if u.username == username:
                user = u
                break
        
        if not user:
            print(f"❌ User {username} not found")
            return None
        
        # Create session
        session_id = self.session_manager.create_session(
            user=user,
            ip_address=ip_address,
            user_agent="Demo Client/1.0"
        )
        
        if session_id:
            self.active_sessions[session_id] = user
            print(f"🔐 User {username} logged in with session {session_id}")
        
        return session_id
    
    def logout_user(self, session_id: str) -> bool:
        """Simulate user logout."""
        if session_id in self.active_sessions:
            user = self.active_sessions[session_id]
            if self.session_manager.delete_session(session_id):
                del self.active_sessions[session_id]
                print(f"🚪 User {user.username} logged out")
                return True
        
        print(f"❌ Session {session_id} not found")
        return False
    
    def session_activity(self, session_id: str, activity_data: Dict[str, Any]) -> bool:
        """Simulate session activity."""
        if session_id not in self.active_sessions:
            print(f"❌ Session {session_id} not active")
            return False
        
        # Update session with activity data
        success = self.session_manager.update_session(session_id, activity_data)
        if success:
            print(f"📝 Session {session_id} activity updated")
        
        return success
    
    def demo_multiple_sessions(self):
        """Demonstrate multiple session management."""
        print("\n👥 Multiple Session Management Demo")
        print("-" * 50)
        
        # Login multiple users
        sessions = []
        for username in ["alice", "bob", "charlie"]:
            session_id = self.login_user(username)
            if session_id:
                sessions.append(session_id)
        
        # Simulate activity
        for i, session_id in enumerate(sessions):
            activity = {
                "page_views": i + 1,
                "last_page": f"/dashboard/{i}",
                "actions": ["view", "edit", "save"][:i+1]
            }
            self.session_activity(session_id, activity)
        
        # Show session stats
        stats = self.session_manager.get_session_stats()
        print(f"\n📊 Session Statistics:")
        for key, value in stats.items():
            print(f"  {key}: {value}")
        
        # Show user sessions
        for user_id in ["user1", "user2"]:
            user_sessions = self.session_manager.get_user_sessions(user_id)
            print(f"\n👤 Sessions for {user_id}:")
            for session in user_sessions:
                print(f"  - {session.session_id}: {session.username} (last: {session.last_activity})")
        
        return sessions
    
    def demo_session_cleanup(self):
        """Demonstrate session cleanup."""
        print("\n🧹 Session Cleanup Demo")
        print("-" * 50)
        
        # Show current active sessions
        active_count = self.session_manager.get_active_sessions_count()
        print(f"Active sessions before cleanup: {active_count}")
        
        # Clean up expired sessions
        expired_count = self.session_manager.cleanup_expired_sessions()
        
        # Show after cleanup
        active_count_after = self.session_manager.get_active_sessions_count()
        print(f"Active sessions after cleanup: {active_count_after}")
        
        return expired_count

def main():
    """Main function demonstrating Redis session management."""
    print("🔐 Redis Session Management Example (Python)")
    print("=" * 60)
    
    try:
        # Initialize demo
        demo = SessionDemo()
        
        # Demo multiple sessions
        sessions = demo.demo_multiple_sessions()
        
        # Demo session cleanup
        demo.demo_session_cleanup()
        
        # Interactive session management
        print("\n" + "=" * 60)
        print("🎯 Interactive Session Management")
        print("Commands:")
        print("  'login <username>' - Login user")
        print("  'logout <session_id>' - Logout session")
        print("  'get <session_id>' - Get session info")
        print("  'update <session_id> <key> <value>' - Update session data")
        print("  'user <user_id>' - Show user sessions")
        print("  'stats' - Show session statistics")
        print("  'cleanup' - Clean up expired sessions")
        print("  'clear' - Clear all sessions")
        print("  'quit' - Exit")
        print("-" * 60)
        
        while True:
            user_input = input("\n📱 Session> ").strip()
            
            if user_input.lower() == 'quit':
                print("👋 Goodbye!")
                break
            elif user_input.lower() == 'stats':
                stats = demo.session_manager.get_session_stats()
                print("📊 Session Statistics:")
                for key, value in stats.items():
                    print(f"  {key}: {value}")
            elif user_input.lower() == 'cleanup':
                expired = demo.demo_session_cleanup()
                print(f"🧹 Cleaned up {expired} expired sessions")
            elif user_input.lower() == 'clear':
                # Clear all active sessions
                for session_id in list(demo.active_sessions.keys()):
                    demo.logout_user(session_id)
                print("🗑️ All sessions cleared")
            elif user_input.startswith('login '):
                username = user_input.split(' ', 1)[1]
                session_id = demo.login_user(username)
                if session_id:
                    print(f"✅ Login successful: {session_id}")
            elif user_input.startswith('logout '):
                session_id = user_input.split(' ', 1)[1]
                success = demo.logout_user(session_id)
                if success:
                    print("✅ Logout successful")
            elif user_input.startswith('get '):
                session_id = user_input.split(' ', 1)[1]
                session = demo.session_manager.get_session(session_id)
                if session:
                    print(f"📄 Session: {session.username} (last: {session.last_activity})")
                    if session.data:
                        print(f"  Data: {session.data}")
                else:
                    print("❌ Session not found")
            elif user_input.startswith('update '):
                parts = user_input.split(' ', 3)
                if len(parts) >= 4:
                    session_id, key, value = parts[1], parts[2], parts[3]
                    success = demo.session_manager.update_session(session_id, {key: value})
                    if success:
                        print("✅ Session updated")
                else:
                    print("Usage: update <session_id> <key> <value>")
            elif user_input.startswith('user '):
                user_id = user_input.split(' ', 1)[1]
                sessions = demo.session_manager.get_user_sessions(user_id)
                if sessions:
                    print(f"👤 Sessions for {user_id}:")
                    for session in sessions:
                        print(f"  - {session.session_id}: {session.username}")
                else:
                    print(f"❌ No sessions found for {user_id}")
            elif user_input:
                print("Unknown command. Type 'quit' to exit.")
        
    except KeyboardInterrupt:
        print("\n\n👋 Demo interrupted by user")
    except Exception as e:
        print(f"\n❌ Unexpected error: {str(e)}")
        sys.exit(1)

if __name__ == "__main__":
    main()